### ISSUE TITLE
Explain how to disable code splitting / chunks generation

### ISSUE THREAD

**Issue by vladshcherbin** [rollup/rollup#2756]
Documentation Is:

- [x] Missing
- [x] Needed
- [ ] Confusing
- [ ] Not Sure?

### Please Explain in Detail...

```js
input: [
  'src/server/one.js',
  'src/server/two.js'
]
```

The code above will create chunks when both files have same imports. Sometimes it's not desired, for example when each file needs to be self-contained (serverless). I searched docs, google, issues, chat and didn't find a way how to turn code splitting off.

Still don't know how solve this. Using array in rollup config is not desired.

### Your Proposal for Changes

It would be nice to see how it's possible to turn off code splitting in the docs.

**Comment by lukastaegert** [comment]
This is not a documentation issue. It is not possible to build several overlapping targets in the same build without code-splitting. One core principle in Rollup's architecture is at the moment that each module exists only once.
Thus an array is the only option for now.

I see however the benefit of being able to do that, which would be performance: As we are sharing the same input options, shared modules do not need to be loaded, transformed or parsed again. In the future, we might consider adding an option to mark entry points as "standalone" to the effect that those entry points are handled in a separate build phase that shares the same parsed modules.

**Comment by vladshcherbin** [comment]
@lukastaegert yeah, this turned out to cause a lot of difficulties for me.

For server code, I don't need chunks. Array in config makes it hard to work with plugins as they run multiple times with different context.

Same for browser, I can't have an `iife` build with an object with 2 entries as `input` because it gives error:

```
Error: UMD and IIFE output formats are not supported for code-splitting builds
```

Don't want to go back to webpack but this situation makes me so sad 😞

*\*changed issue to a feature request*

**Comment by shellscape** [comment]
@lukastaegert thoughts on the state of this one?

**Comment by lukastaegert** [comment]
At the moment this is lower priority. I would prefer to put more work in a cache plugin API that could be utilised by the CLI to share information between builds in an array of inputs. If this is done, it could be possible to reintroduce this as a CLI feature.

**Comment by cref** [comment]
IMHO it would be more intuitive to simply disable the code-splitting feature for module formats for which this feature isn't supported. Maybe with a warning or by requiring a force flag?

**Comment by alshdavid** [comment]
When trying to build a `umd` formatted library for npm with `preserveModules: true` I am greeted with 

```
Error: UMD and IIFE output formats are not supported for code-splitting builds
```

Obviously, this is a library for npm so I don't want or need code splitting in this context - I just need maximum import compatibility.

```typescript
export default {
  input: `src/index.ts`,
  preserveModules: true,
  output: {
    format: 'umd',
    dir: './dist'
  },
  external: [],
  watch: {
    include: 'src/**',
  },
  plugins: [
    json(),
    typescript({ useTsconfigDeclarationDir: true }),
    commonjs(),
    resolve(),
    sourceMaps(),
  ],
}
```

**Comment by lukastaegert** [comment]
The whole point of `preserveModules` is to generate multiple chunks and is inherently incompatible with a UMD of IIFE output. If you do not need it, do not use it.

**Comment by alshdavid** [comment]
You're right, I mistakenly thought webpack wouldn't read import/export syntax in node_modules

**Comment by webpepper** [comment]
I've just hit the same roadblock. I'm building a serverless application in which I want standalone files (bundles) in the end. In my case I don't want any imports, I want all the required code in the same file: one source file = one completely standalone compiled file. X input files = X outputs files. Currently it seems that with multiple inputs the bundler... doesn't actually bundle _everything_.

A semi-working solution is to export multiple configs from rollup config, i.e. `module.exports = [{}, {}, {}]` but it gets slower and uses crazy amounts of memory with each additional input. It worked great in the beginning when there were a couple of input files, but now the build process is very slow and crashes at times even when running with node's `--max-old-space-size` flag.

Another solution is to put all these shared modules (output chunks) into a separate folder and include them with every single serverless function. You end up including unneeded code in your serverless functions. So this is a suboptimal solution too.
```js
{
    inputs: [ { function1: 'src/function1.js', function2: 'src/function2.js' } ],
    output: {
       entryFileNames: '[name].js',
       chunkFileNames: 'includes/[hash].js',
    }
}
```
When deploying you have to copy this entire `includes` (or whatever you name it) folder with every function because you don't know which files exactly will be needed.

**Comment by skeggse** [comment]
See also #2398.

**Comment by Geo1088** [comment]
This is making it difficult to use Rollup for my browser extension. I want to use Rollup because I'm working around a bug that prevents me from using ES6 modules natively in Firefox extension content scripts; the ability to use `format: 'iife'` to convert modules to IIFEs has enabled me to use ES6 modules for code organization despite the lack of support from Firefox.

However, if I want to build multiple files (e.g. different content scripts that operate on separate domains) by passing an array to `input`, I run into `Error: UMD and IIFE output formats are not supported for code-splitting builds`. In order to get the result I want, I'm forced to use multiple build configurations, which is not ideal.

I realize that disabling code splitting still wouldn't be a perfect solution, because shared code will be duplicated between the generated bundles, but it seems like it'd be the best way for my use-case - other than maybe generating an `es` build, converting all the individual chunks to `iife`, and then adding the appropriate chunks to the appropriate locations in my manifest.

**Comment by rjb3977** [comment]
Also would like to see this made possible. It'd let me get around Firefox not supporting `{ type: 'module' }` for `Worker`s. I feel like a configuration option for it named something like `output.standalone` would make sense. Even if #2398 gets implemented, that would only really be a workaround for this issue. There are cases where you can't split everything into separate configurations, such as if you have a plugin that creates chunks via `emitFile` that you also want to be standalone.

**Comment by richardtallent** [comment]
I'm having the same issue. I'm using Vite, and I need to ensure that **every** ounce of JavaScript ends up in a single file, with no `preload` links or internal `import` statements, so I can create a single-file distributable HTML file that will load from the desktop.

This was working awhile back for Vite, but then a change was made there that is now resulting in a `vendor` file being split off, and I don't seem to have any control over that happening.

**Comment by HongSiyuan** [comment]
> I'm having the same issue. I'm using Vite, and I need to ensure that **every** ounce of JavaScript ends up in a single file, with no `preload` links or internal `import` statements, so I can create a single-file distributable HTML file that will load from the desktop.
> 
> This was working awhile back for Vite, but then a change was made there that is now resulting in a `vendor` file being split off, and I don't seem to have any control over that happening.

@richardtallent you can try this in the vite config file. It works for me.
```js
build: {
  rollupOptions: {
    output: {
      manualChunks: {}
    }
  }
}
```

**Comment by richardtallent** [comment]
Hi @HongSiyuan,

Thanks! Yes, we were able to update [vite-plugin-singlefile](https://github.com/richardtallent/vite-plugin-singlefile) with a similar option for `manualChunks` (setting it to a function returning a single name works too).

**Comment by xenobytezero** [comment]
I also have a use-case where I am bundling a Web Component that is dynamically imported on the client ( using `import()` )  and so needs the component and all of its dependencies in a single file that can be downloaded in one go.

Ideally I would use IIFE or UMD, but I get the "not compatible with code-splitting" error, even though I'm explicitly not wanting to code split. Is there a way I can see what's trying to be code split so I can work around it?

I've spent most of the week touring all of the bundlers trying to get away from Webpack, and Rollup by far is the most extendable I've seen so far. I can't switch all our stuff over to it unless it has this feature.

**Comment by shadow-light** [comment]
> Hi @HongSiyuan,
> 
> Thanks! Yes, we were able to update [vite-plugin-singlefile](https://github.com/richardtallent/vite-plugin-singlefile) with a similar option for `manualChunks` (setting it to a function returning a single name works too).

Nice find. Setting `manualChunks` to `undefined` and `{}` stopped Vite creating a vendor module but didn't force dynamic imports to be inlined, where as for some reason `() => 'everything.js'` did.

**Comment by wensonsmith** [comment]
I'm building a chrome extension based on [antfu/vitesse-webext](https://github.com/antfu/vitesse-webext) now

The extension has two content scripts for different website that has same dependency .

my build config 

```js
    rollupOptions: {
      input: {
        a: r('src/contentScripts/a.ts'),
        b: r('src/contentScripts/b.ts'),
      },
      output: {
        entryFileNames({ name }) {
          return `${name}.global.js`
        },
      },
    },
```

rollup always chunk the shared dependency , so content script can't work because it `import` the chunk after build 

```js
#a.global.js
import {s as e} from"./chunk.js";
```

how to disable chunk and build two standalone js file , is there some hacking solution ?

@lukastaegert  I hope to raise the priority because it has been two years ..   thanks

**Comment by manlu0x01** [comment]
It is highly recommended to raise the priority of this feature.

We face the same problem when building a self-supporting component library using css-in-js. rollup packages the public code into the assets directory by default, so that individual components cannot be atomized.

**Comment by fdrobidoux** [comment]
This is also a problem we're having with the way we're building our Vue 3 app with Vite. Can this issue get some love, please?

**Comment by bigskillet** [comment]
I'm having the same problem and would also like to see this feature implemented.

**Comment by eritbh** [comment]
For web extensions specifically, I created a plugin that can take a `manifest.json` as a single input and run each of its entry points through Rollup individually, rewriting the manifest in the process. The whole thing is a bit of a hack and could use a lot of work, but it might be useful as a starting point for others working on browser extensions.

I don't think this plugin approach is ideal as a permanent solution to this issue (a full solution of this type would ideally handle non-code entry points like icons and `webAccessibleResources`, but at that point it feels like it'd be better suited to a Webpack config instead of Rollup), but in the absence of anything better, I figured I'd share it in case it's useful to anyone else.

**Comment by KashubaK** [comment]
In my case, I'm building both an app and a website plugin in the same project. The app is a form builder, and the website plugin is the form logic that can be embedded into a website. There are two main entrypoints: `application.tsx`, and `embed.tsx`. 

My problem is that since the app also depends on the embedded form logic, it creates a separate chunk called something like `Form.91e05067.js`. This is all fine and dandy for the application because it can just request `/Form.91e05067.js` from the same host and it works. However this gets complicated for the website plugin, as it's much simpler to just offer a single JS file for the end-user's website. 

For a more realistic idea, this is a Shopify app that directly integrates into our users' themes. It's much easier for our auto-installation process to upload a single `embed.js` file and have their theme request it as opposed to uploading all of the different chunks and clean up the old ones.

We are currently using Webpack to handle this and it works as expected, however we want to transition to Vite. Unfortunately this is holding us up.

**Comment by rkyoku** [comment]
+1 here!

Needing to generate a single fat output file with all imports inlined for each input/entry file (my rollup config has multiple inputs).

I no longer want code-splitting because it is badly rated by Lighthouse, which complains about performance issues (delay before loading all import-ed scripts, and many many many HTTP requests instead of a single one).

Using a function for the `manualChunks` option does not work (I have the first output file with all the webapp code in it, and every other output file is importing this one).

Using an empty object/hash for this option does not work either ==> imports are not resolved.

I will try to see if I can `export default` an array of config parameters instead of a single object, and if so I will try to map over my input files and use `inlineDynamicImports` option (which I cannot use otherwise, as it does not support multiple inputs)

@lukastaegert what would be the solution to this seemingly common issue? (one single fat output file for each input file)

**Comment by lukastaegert** [comment]
Use an array of inputs, one for each input file.

**Comment by rkyoku** [comment]
Hi @lukastaegert ,

Thank you for the reply, and this is exactly what I ended up doing.

However, as someone already suggested, a simple `use-code-splitting` option (defaulting to `true`) would be easy to add, and surely make things way easier for people.

If it may help someone here, here is my code for achieving a "single big fat output file" for each input file:

```js
const INPUT_DIR = 'app/webapp/asset/js/page/';
const OUTPUT_DIR = 'public/webapp/js/iife';

return fs
	.readdirSync(INPUT_DIR)
	.filter(file => file.endsWith('.js'))
	.map(file => ({
		input: INPUT_DIR + file,
		external: [/* some externals */],
		output: [{
				inlineDynamicImports: true,
				entryFileNames: `${file}.${version}.js`,
				dir: OUTPUT_DIR,
				format: 'iife',
				sourcemap: false,
				globals: {/* some globals */},
				plugins: [/* separate dev/dist plugins! */]
		}]
	}))
;
```

**Comment by lukastaegert** [comment]
It could be a CLI-only option, but for Rollup core, it is actually quite difficult as everything is designed around the concept "each module exists only once in the graph". This does not mean that we might not add a way to support this eventually in core as there would likely be some performance benefits, but it will be a major architectural change.

**Comment by antonreshetov** [comment]
Any news?

**Comment by stephdz** [comment]
We need it ! Not everyone builds an SPA !

My use case is multiple entries and get one single JS bundle for each. Vite cannot do that due to Rollup not able to :(

**Comment by Yash-Singh1** [comment]
Any updates on this?

**Comment by softwareCobbler** [comment]
With Vite, setting `output.manualChunks = () => "app"` seems to produce a single js and a single css file, where without the option it would generate 100 different chunks, but the name of the chunks aren't "app.js"/"app.css", but rather "index.js"/"index.css". Does it discard the function return value?

**Comment by devmuch** [comment]
> With Vite, setting `output.manualChunks = () => "app"` seems to produce a single js and a single css file, where without the option it would generate 100 different chunks (or however many), but the name of the chunks aren't "app.js"/"app.css", but rather "index.js"/"index.css". Does it discard the function return value?

Mine returns the name of the first key in `input` object and ignores the function output string.

```
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
import { resolve } from "path";

export default defineConfig({
  plugins: [react()],
  build: {
    rollupOptions: {
      input: {
        folder-a: resolve(
          __dirname,
          "folder-a/index.html"
        ),
        folder-b: resolve(
          __dirname,
          "folder-b/index.html"
        ),
      },
      output: {
        manualChunks: (id) => {
          if (/\/react-dom\//.test(id)) return "react-dom";
          if (/\/react\//.test(id)) return "react";
          return "app";
        },
        entryFileNames: `assets/[name].js`,
        assetFileNames: `assets/[name].[ext]`,
        chunkFileNames: "assets/[name]-chunk.js",
      },
    },
  },
});
```
would resolve:
```
dist/assets/react-chunk.js 
dist/assets/folder-a.css 
dist/assets/react-dom-chunk.js 
dist/assets/folder-a.js
```

**Comment by lmanzke** [comment]
Same use-case here. Multiple entries, multiple independent js bundles needed. Any updates?

**Comment by clementmas** [comment]
@softwareCobbler's solution worked for me. Adding this to `vite.config.js`:

```
// Disable code splitting
build: {
    rollupOptions: {
        output: {
            manualChunks: () => 'app',
        },
    },
},
```
Produces a single app.js (and app.css) file instead of many.

**Comment by MatthewSbar** [comment]
In case anyone else wanted to switch from create-react-app to vite while keeping the "one js file" output, you can use the above trick, but if you want your single js chunk to just be named `index.js` rather than `index.{chunkname}.js`, you can change your build command in vite like so

```
// package.json
scripts: {
"build": "tsc && vite build && mv $(find dist/assets -name 'index.*.js') dist/assets/index.js",
}
```

**Comment by sidneifjr** [comment]
> @softwareCobbler's solution worked for me. Adding this to `vite.config.js`:
> 
> ```
> // Disable code splitting
> build: {
>     rollupOptions: {
>         output: {
>             manualChunks: () => 'app',
>         },
>     },
> },
> ```
> 
> Produces a single app.js (and app.css) file instead of many.

This solved my issue, thank you very much!

**Comment by jiby-aurum** [comment]
Same use case here, building lambda functions. 

What I ended up doing was generating the chunked build with all transformations in a first pass, and then doing multiple build (one per lambda function) without any transformations to join up the chunks. Ended up being much faster than just doing multiple builds directly on the source.

**Comment by aarjithn** [comment]
> @shadow-light solution works:
> 
> ```js
> export default {
>     input: "src/lips.js",
>     output: {
>         name: "lips",
>         file: "dist/lips.js",
>         format: "umd",
>         globals: {
>             "@babel/runtime/regenerator": "regeneratorRuntime"
>         },
>         manualChunks: () => 'everything.js'
>     },
>     ...
> };
> ```
> 
> I think that the name the function returns doesn't matter.

This doesn’t seem to work for multiple entries where you wish to have one standalone output per entry.

**Comment by hood** [comment]
> > @shadow-light solution works:
> > ```js
> > export default {
> >     input: "src/lips.js",
> >     output: {
> >         name: "lips",
> >         file: "dist/lips.js",
> >         format: "umd",
> >         globals: {
> >             "@babel/runtime/regenerator": "regeneratorRuntime"
> >         },
> >         manualChunks: () => 'everything.js'
> >     },
> >     ...
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think that the name the function returns doesn't matter.
> 
> This doesn’t seem to work for multiple entries where you wish to have one standalone output per entry.

Can confirm here, having the following as my `vite.config.ts` file:
```ts
export default defineConfig({
  build: {
    minify: false,
    rollupOptions: {
      input: [
        "./src/one/One.svelte",
        "./src/one/Child.svelte",
        "./src/one/one-entry.js",

        "./src/two/Two.svelte",
        "./src/two/Child.svelte",
        "./src/two/two-entry.js",
      ],
      output: {
        format: "esm",
        compact: true,
        entryFileNames: "[name].js",
      },
    },
  },
  plugins: [
    svelte({
      preprocess: SveltePreprocess({ scss: true, typescript: true }),
      compilerOptions: {
        dev: false,
        customElement: false,
      },
    }),
  ],
});
```

I can't get a simple `One.svelte` -> `One.js` export once I get different components depending on the same stuff. 

The `manualChunk` option doesn't work in my scenario because, as @rkyoku said already, then all modules end up being transpiled to one-line js files importing one of the others which gets randomly picked to hold all the transpilation output.

Is there really no roadmap for this feature?

**Comment by anisg** [comment]
> I'm building a chrome extension based on [antfu/vitesse-webext](https://github.com/antfu/vitesse-webext) now
> 
> The extension has two content scripts for different website that has same dependency .
> 
> my build config
> 
> ```js
>     rollupOptions: {
>       input: {
>         a: r('src/contentScripts/a.ts'),
>         b: r('src/contentScripts/b.ts'),
>       },
>       output: {
>         entryFileNames({ name }) {
>           return `${name}.global.js`
>         },
>       },
>     },
> ```
> 
> rollup always chunk the shared dependency , so content script can't work because it `import` the chunk after build
> 
> ```js
> #a.global.js
> import {s as e} from"./chunk.js";
> ```
> 
> how to disable chunk and build two standalone js file , is there some hacking solution ?
> 
> @lukastaegert I hope to raise the priority because it has been two years .. thanks

Quick hack for this: use npm alias.

let's say both use "@popperjs/core@2.11.7"

```
yarn add popperjs-content-script1@npm:@popperjs/core@2.11.7
yarn add popperjs-content-script2@npm:@popperjs/core@2.11.7
```

then in content script 1:
```
import {} from "popperjs-content-script1"
```

in content script 2:
```
import {} from "popperjs-content-script2"
```

**Comment by yashspr** [comment]
**The solution below is only for those using Vite.** I have multiple input files and I wanted multiple outputs to be generated which are self contained i.e. no dependencies should be shared between the output chunks. For this, the above solutions didn't work. The workaround I managed is to run the build step using the vite javascript API for each input separately. This generated separate self contained files for each of the entry points.

```
import { fileURLToPath } from 'url'
import { build } from 'vite'
import rimraf from 'rimraf'
import preact from '@preact/preset-vite'
import cssInjectedByJsPlugin from 'vite-plugin-css-injected-by-js'

const __dirname = fileURLToPath(new URL('.', import.meta.url))

const packages = [
	{ pk1: fileURLToPath(new URL('./packages/pkg1/main.jsx', import.meta.url)) }
]

async function buildPackages() {
	await rimraf("./dist");

	for (const _package of packages) {
		await build({
			publicDir: false,
			plugins: [
				preact(),
				cssInjectedByJsPlugin()
			],
			build: {
				rollupOptions: {
					input: _package,
					output: {
						entryFileNames: "js/[name].js",
					},
				},
				emptyOutDir: false
			},
			configFile: false,
		})
	}
}

buildPackages();
```

Just create a `build.js` file at the root. And run this with node.

**Comment by jim-spot** [comment]
> Use an array of configs, one for each input file. Or in other words, run rollup once per input file.

Not sure why the downvotes here, I actually think this is the best answer. Here is an example of how I can use this in a rollup.config.js file to output three separate files that don't get code-split:


```javascript
const defaultConfig = {
    output: {
        dir: "./build",
        format: "cjs"
    },
    plugins: [
        commonjs(),
        resolve({
            browser: false,
            preferBuiltins: true,
            exportConditions: ["node"]
        }),
        json(),
        typescript({
            compilerOptions: {
                module: "esnext",
                target: "es2020",
                outDir: undefined
            }
        }),
        optimizeLodashImports(),
        terser()
    ]
};

export default [
    {
        input: "./handlers/PTZSFTPFHandler.ts",
        ...defaultConfig
    },
    {
        input: "./handlers/KinesisConsumerHandler.ts",
        ...defaultConfig
    },
    {
        input: "./handlers/SQSWorkerHandler.ts",
        ...defaultConfig
    }
];
```

**Comment by woss** [comment]
@jim-spot if you don't mind me making an observation. `spread`ing the object AFTER the customization is quite bug prone and somewhat antipattern. what you want is to have default -> overrides, so you example should be 
```
 ...
    {
        ...defaultConfig,
        input: "./handlers/PTZSFTPFHandler.ts"
    },
...
```

in this case you can have a default state and still have the overrides.

**Comment by jim-spot** [comment]
> @jim-spot if you don't mind me making an observation. `spread`ing the object AFTER the customization is quite bug-prone and somewhat antipattern. what you want is to have default -> overrides, so your example should be
> 
> ```
>  ...
>     {
>         ...defaultConfig,
>         input: "./handlers/PTZSFTPFHandler.ts"
>     },
> ...
> ```
> 
> in this case you can have a default state and still have the overrides.

Actually I don't view it as anti-pattern at all. It all depends on whether or not you consider the spread object to be a hard override or a default. In this case it doesn't matter since the object is well-defined up top. I could have gone either way in this case.

Agreed that it **does** matter in many cases if you view it as an override or a default, but to say that one is anti-pattern doesn't seem right.

**Comment by nicholasdavidbrown** [comment]
Just to add on to @yashspr's answer for those using Vite, you can use ```glob``` to select input files directly, rather than keeping an updated list of your files as a variable.

I also disabled minification. This was a unique requirement for a legacy system running with old JavaScript.

This runs a build script for each file referenced, where you'll need to install the modules if you don't already use them. 

```
import { build } from 'vite';
import { glob } from 'glob';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { rimraf } from 'rimraf';

async function buildPackages () {
  await rimraf('./dist');

  const inputFiles = Object.fromEntries(
    glob.sync('src/**/*.js').map(file => [
      // This remove `src/` as well as the file extension from each
      // file, so e.g. src/nested/foo.js becomes nested/foo
      path.relative(
        'src',
        file.slice(0, file.length - path.extname(file).length)
      ),
      // This expands the relative paths to absolute paths, so e.g.
      // src/nested/foo becomes /project/src/nested/foo.js
      fileURLToPath(new URL(file, import.meta.url))
    ])
  );

  for (const key in inputFiles) {
    const inputFile = inputFiles[key];
    console.log('Building', inputFile);
    await build({
      publicDir: false,
      build: {
        minify: false,
        // sourcemap: true,
        outDir: 'dist',
        rollupOptions: {
          input: inputFile,
          output: {
            entryFileNames: 'scripting/[name].js',
            globals: {
              xml: 'xml',
              newxml: 'newxml',
             // ... etc
            }
          }
        },
        emptyOutDir: false
      },
      configFile: false
    });
  }
}

buildPackages();
```

Saved as the ```build.js``` file and used instead of the default build script via node ```node build.js```.

**Comment by thescientist13** [comment]
Just wanted to echo others who have commented that a feature like this would be really useful for the serverless deployment target space, as that is the exact use case that got me to this issue.  Seems like there are some work-arounds to explore in the meantime so will give those a shot.  👍

**Comment by jim-spot** [comment]
> Just wanted to echo others who have commented that a feature like this would be really useful for the serverless deployment target space, as that is the exact use case that got me to this issue. Seems like there are some work-arounds to explore in the meantime so will give those a shot. 👍

I am specifically using this for Serverless deployments where I need to compile three separate bundles of code for three separate functions. The multi-config solution mentioned above works perfectly and I would argue it's not really a work-around— it IS the supported way in rollup and no new features is needed. If any change needs to be made, it's probably that the documentation should be updated to reflect this.

**Comment by thescientist13** [comment]
Good call out thanks @jim-spot .  Yeah, shouldn't be much work at all to implement it as you say.  Will give it a try soon.  👍

**Comment by webarthur** [comment]
I find a way to split scripts, one for a main.js scritp and another script for each page:

```javascript
import { basename } from 'path'
// import rimraf from 'rimraf'
import { build } from 'vite'
import glob from 'glob'
import { fileURLToPath, URL } from 'url'

import includeHTMLPlugin from './plugins/vite-plugin-include-html.js'

const __dirname = fileURLToPath(new URL('.', import.meta.url))

const pages = glob.sync('./**/*.html', { 
  ignore:[
    './node_modules/**',
    './partials/**',
    './dist/**',
  ]
}).map(file => [
  // Remove the extension to create a JavaScript file with the name of the page
  basename(file).replace(/\.[^.]+$/, ''),
  // This expands the relative paths to absolute paths, so e.g.
  // src/nested/foo becomes /home/username/dev/projectname/src/nested/foo.js
  fileURLToPath(new URL(file, import.meta.url))
])

// console.log(pages)

async function buildPackages() {
	// await rimraf("./dist")

    for (const [filename, filepath] of pages) {
    console.log('Building', filepath)

		await build({
			publicDir: false,
			plugins: [
				includeHTMLPlugin(`${ __dirname }/partials/`)
			],
			build: {
				rollupOptions: {
					input: {
            [filename]: filepath, // Page script
            main: 'main.js',      // Shared main file
          },
					output: {
						entryFileNames: "js/[name]-[hash].js",
					},
				},
				emptyOutDir: false
			},
			configFile: false,
		})
	}
}

buildPackages()
```

Hope it helps somehow.

**Comment by intelcoder** [comment]
Removing code splitting in dev mode does not seem to work. Is there any other solution for this? It fetches like 10 js chunks in dev mode but not in production mode.

**Comment by supermacro** [comment]
Related issue: https://github.com/rollup/rollup/issues/5212

**Comment by toFrankie** [comment]
Releted issue: [vite #1683](https://github.com/vitejs/vite/issues/1683)

**Comment by AdamAnSubtractM** [comment]
Hey team, any updates on this? Is this likely ever to be a thing/on the roadmap? Anyone have success achieving this by using other build tools? Thanks in advance.

**Comment by supermacro** [comment]
> Hey team, any updates on this? Is this likely ever to be a thing/on the roadmap? Anyone have success achieving this by using other build tools? Thanks in advance.

fwiw I moved away from vite in favor of bun, and was able to achieve this by way of having 2 separate builds and implementing the build programmatically:

![image](https://github.com/rollup/rollup/assets/10621548/4b5a5d51-bd5d-4b66-bb4e-8cdfcd0e77eb)

**Comment by rahulbhanushali** [comment]
When trying to build in lib mode - @jim-spot's [solution above](https://github.com/rollup/rollup/issues/2756#issuecomment-1572799589) doesn't work with vite. When exporting the array, build fails with below error:
```
error during build:
Error: config must export or return an object.
    at loadConfigFromFile (file:///..../node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:68013:19)
    at async resolveConfig (file:///.../node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:67619:28)
    at async build (file:///.../node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:66762:20)
    at async CAC.<anonymous> (file:///.../node_modules/vite/dist/node/cli.js:842:9)
```

I was able to tweak the solutions proposed by @yashspr and @nicholasdavidbrown for lib mode by create a `build.js` as follows:
```
import { defineConfig } from 'vite';
import { resolve } from 'path';
import preact from '@preact/preset-vite';
import { viteStaticCopy } from 'vite-plugin-static-copy';
import { build } from 'vite';
import { fileURLToPath } from 'url';

// https://vitejs.dev/config/
const __dirname = fileURLToPath(new URL('.', import.meta.url));

const inputFiles = [
   {
      outputFileName: 'index',
      inputFilePath: 'index.js',
      emptyOutDir: true,
      staticTargets: [
         { src: resolve(__dirname, 'src/assets/fonts/*'), dest: 'fonts' },
         { src: resolve(__dirname, 'src/assets/icons/*'), dest: 'icons' },
      ],
   },
   {
      outputFileName: 'widget1',
      inputFilePath: 'src/lib/widget1.jsx',
   },
   {
      outputFileName: 'widget2',
      inputFilePath: 'src/lib/widget2.jsx',
   },
   {
      outputFileName: 'DownloadAssetWidget',
      inputFilePath: 'src/lib/DownloadAssetWidget/index.jsx',
   },
   {
      outputFileName: 'SurveyWidget',
      inputFilePath: 'src/lib/SurveyWidget/index.jsx',
   },
];

const buildConfig = (fileConfig) => {
   let plugins = [preact()];

   if (fileConfig.staticTargets) {
      plugins.push(
         viteStaticCopy({
            targets: fileConfig.staticTargets,
            hook: 'writeBundle', // run the plugin after vite build
         }),
      );
   }

   return defineConfig({
      configFile: false,
      build: {
         emptyOutDir: fileConfig.emptyOutDir == true,
         lib: {
            entry: {
               [fileConfig.outputFileName]: resolve(__dirname, fileConfig.inputFilePath),
            },
            formats: ['es'],
         },
      },
      plugins: plugins,
   });
};

const finalBuild = inputFiles.map((fileConfig) => buildConfig(fileConfig));

async function buildPackages() {
   for (const config of finalBuild) {
      await build(config);
   }
}

// export default finaBuild;
export default buildPackages();

```

Simple run `node build.js` to generate the build output in dist.

**Comment by jim-spot** [comment]
@rahulbhanushali any chance you might share the config that didn't work for you? Also would be curious to know the version you were on. We've been using that config without fail on many projects now.

**Comment by wuyuweixin** [comment]
Any progress ?

**Comment by VincEnterprise** [comment]
> When trying to build in lib mode - @jim-spot's [solution above](https://github.com/rollup/rollup/issues/2756#issuecomment-1572799589) doesn't work with vite. When exporting the array, build fails with below error:
> 
> ```
> error during build:
> Error: config must export or return an object.
>     at loadConfigFromFile (file:///..../node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:68013:19)
>     at async resolveConfig (file:///.../node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:67619:28)
>     at async build (file:///.../node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:66762:20)
>     at async CAC.<anonymous> (file:///.../node_modules/vite/dist/node/cli.js:842:9)
> ```
> 
> I was able to tweak the solutions proposed by @yashspr and @nicholasdavidbrown for lib mode by create a `build.js` as follows:
> 
> ```
> import { defineConfig } from 'vite';
> import { resolve } from 'path';
> import preact from '@preact/preset-vite';
> import { viteStaticCopy } from 'vite-plugin-static-copy';
> import { build } from 'vite';
> import { fileURLToPath } from 'url';
> 
> // https://vitejs.dev/config/
> const __dirname = fileURLToPath(new URL('.', import.meta.url));
> 
> const inputFiles = [
>    {
>       outputFileName: 'index',
>       inputFilePath: 'index.js',
>       emptyOutDir: true,
>       staticTargets: [
>          { src: resolve(__dirname, 'src/assets/fonts/*'), dest: 'fonts' },
>          { src: resolve(__dirname, 'src/assets/icons/*'), dest: 'icons' },
>       ],
>    },
>    {
>       outputFileName: 'widget1',
>       inputFilePath: 'src/lib/widget1.jsx',
>    },
>    {
>       outputFileName: 'widget2',
>       inputFilePath: 'src/lib/widget2.jsx',
>    },
>    {
>       outputFileName: 'DownloadAssetWidget',
>       inputFilePath: 'src/lib/DownloadAssetWidget/index.jsx',
>    },
>    {
>       outputFileName: 'SurveyWidget',
>       inputFilePath: 'src/lib/SurveyWidget/index.jsx',
>    },
> ];
> 
> const buildConfig = (fileConfig) => {
>    let plugins = [preact()];
> 
>    if (fileConfig.staticTargets) {
>       plugins.push(
>          viteStaticCopy({
>             targets: fileConfig.staticTargets,
>             hook: 'writeBundle', // run the plugin after vite build
>          }),
>       );
>    }
> 
>    return defineConfig({
>       configFile: false,
>       build: {
>          emptyOutDir: fileConfig.emptyOutDir == true,
>          lib: {
>             entry: {
>                [fileConfig.outputFileName]: resolve(__dirname, fileConfig.inputFilePath),
>             },
>             formats: ['es'],
>          },
>       },
>       plugins: plugins,
>    });
> };
> 
> const finalBuild = inputFiles.map((fileConfig) => buildConfig(fileConfig));
> 
> async function buildPackages() {
>    for (const config of finalBuild) {
>       await build(config);
>    }
> }
> 
> // export default finaBuild;
> export default buildPackages();
> ```
> 
> Simple run `node build.js` to generate the build output in dist.

Has anyone figured out how to use this multiple entry points config but with shared vendor chunks?

So for example:

Widget 1 and Widget 2 both use React, so a build script would generate the following files:

-  `chunk-vendor-all.js`  (includes react)
-  `chunk-widget-1-vendors.js` (includes deps specific only to widget 1)
-  `chunk-widget-2-vendors.js` (includes deps specific only to widget 2)
-  `widget-1.js`
-  `widget-2.js`

I really need this type of config because I'm trying to upgrade from this type of webpack setup:

[https://gist.github.com/Akryum/ece2ca512a1f40d70a1d467566783219 ](https://gist.github.com/Akryum/ece2ca512a1f40d70a1d467566783219 )

**Comment by alyyousuf7** [comment]
I was able to write a small plugin that will disable code-splitting for specific files:

```js
const disableCodeSplitting = (targets = []) => {
  /** @type{import('rollup').Plugin} */
  const plugin = {
    name: 'disable-code-splitting',
    async resolveId(source, importer, options) {
      const resolved = await this.resolve(source, importer, options);
      if (resolved && targets.some(file => resolved.id.includes(file))) {
        return `${resolved.id}?unique=${Math.random()}`;
      }
    },
    load(id) {
      const regex = /(\?unique=.*)$/;
      if (regex.test(id)) {
        return this.load({ id: id.replace(regex, '') });
      }
    },
  };

  return plugin;
};
```

Usage:
```js
// rollup.config.mjs
export default defineConfig([{
  input: ['src/index1.ts', 'src/index2.ts'],
  output: {
    dir: 'dist',
  },
  plugins: [
    disableCodeSplitting(['src/shared.ts']),
    typescript({
      tsconfig: 'tsconfig.json',
      module: 'ESNext',
    }),
  ],
}]);
```

`src/`:
```js
// src/shared.ts
export default { hello: 'world' };

// src/index1.ts
import Shared from './shared';
console.log('index1.ts', Shared);

// src/index2.ts
import Shared from './shared';
console.log('index2.ts', Shared);
```

`dist/`:
```js
// dist/index1.js
var Shared = { hello: 'world' };
console.log('index1.ts', Shared);

// dist/index2.js
var Shared = { hello: 'world' };
console.log('index2.ts', Shared);
```

I hope this helps someone.

I didn't put too much effort into figuring out if its possible to auto detect the targets.

**Comment by ManneJayaSurya** [comment]
My project is webpack currently 
and i am trying movie to vite and i have multiple entry point

while running vitebuild command it creating a chunck 
Is other any supported way to stop creating chunck files or any suggestions 

I already tried by using manualchunck but still chunck file was getting generated

**Comment by JackyM06** [comment]
There has been no progress until 2024. 
I'm trying to implement a plugin that compiles the service worker separately, but I'm currently blocked here ...
```typescript
import { PluginOption } from 'vite';
import path from 'node:path';

/**
 * 生成一段能够插入HTML的Script标签，用于注册Service Worker
 * @param filename
 * @param scope
 * @returns
 */
function generateSWScript(src: string, scope = '/', isModule = false): string {
    return `<script id=service-worker>
                if('serviceWorker' in navigator){ 
                    navigator.serviceWorker.register('${src}', { 
                        scope: '${scope}',
                        ${isModule ? ", type: 'module'" : ''}
                    }) 
                }
            </script>`;
}

/**
 * Service Worker Vite插件，用于打包项目中的单个sw.js
 * @param file sw.ts在项目中的位置
 * @param scope 手动指名的作用域
 * @param option.renderFile 插入到HTML中实际注册的文件地址.
 * @param option.scope 指定sw注册时的作用域
 * @param option.devPort 本地服务器端口
 * @returns
 */
export default function (
    file: string,
    option = {
        output: '',
        renderFile: '',
        scope: '',
    }
): PluginOption {
    const { renderFile, scope, output } = option;
    const sourceFile = renderFile.replace('.js', path.extname(file));

    return {
        name: 'vite-plugin-service-worker',

        config() {
            return {
                build: {
                    rollupOptions: {
                        output: {
                            manualChunks: {
                                [file]: [file],
                            },
                        },
                    },
                },
            };
        },

        async buildStart() {
            if (this.meta.watchMode) {
                return;
            }
            this.emitFile({
                type: 'chunk',
                fileName: output || 'assets/sw.js',
                id: file,
            });
        },

        configureServer(server) {
            server.middlewares.use(async (req, _, next) => {
                if (req.url === sourceFile) {
                    req.url = path.join(server.config.base || '', file);
                }
                next();
            });
        },

        transformIndexHtml(html, ctx): string {
            const command = ctx.server?.config.command;
            const isServe = command === 'serve';
            return html.replace(
                '</body>',
                `  ${generateSWScript(isServe ? sourceFile : renderFile, scope, isServe)}\n  </body>`
            );
        },
    };
}

```

**Comment by ChuanyuWang** [comment]
Any updates? It's really important and required urgently. The thread live actively from 2019 tells anything.

**Comment by raegen** [comment]
> I was able to write a small plugin that will disable code-splitting / chunking a specific files out.
> 
> The idea is the same as [this guy](https://github.com/TrickyPi/github-contributions-colorful/blob/fe3090f4428a850e53752410aedf8e35892d12c9/vite.config.ts#L25-L43), but with some better support for any file type (hopefully?) that your existing config can parse.
> 
> ```js
> import path from 'path';
> 
> const disableChunks = (targets = []) => {
>   /** @type{import('rollup').Plugin} */
>   const plugin = {
>     name: 'disable-chunks',
>     async resolveId(source, importer, options) {
>       const resolved = await this.resolve(source, importer, options);
>       if (resolved && targets.some(file => resolved.id.includes(file))) {
>         return `${resolved.id}?unique=${Math.random()}`;
>       }
>     },
>     load(id) {
>       const regex = /(\?unique=.*)$/;
>       if (regex.test(id)) {
>         return this.load({ id: id.replace(regex, '') });
>       }
>     },
>    generateBundle(options, bundle) {
>       const nonEntryChunks = Object.values(bundle).filter(({ isEntry }) => !isEntry);
>       if (nonEntryChunks.length > 0) {
>         const moduleIds = new Set(
>           nonEntryChunks.map(({ moduleIds }) => moduleIds).flat()
>             .map(absPath => path.relative('.', absPath)),
>         );
>         throw new Error(`Found non-entry chunks: ${Array.from(moduleIds).join(', ')}`);
>       }
>     },
>   };
> 
>   return plugin;
> };
> ```
> 
> Usage:
> 
> ```js
> // rollup.config.mjs
> export default defineConfig([{
>   input: ['src/index1.ts', 'src/index2.ts'],
>   output: {
>     dir: 'dist',
>   },
>   plugins: [
>     disableChunks(['src/shared.ts']),
>     typescript({ module: 'ESNext' }),
>   ],
> }]);
> ```
> 
> `src/`:
> 
> ```js
> // src/shared.ts
> export default { hello: 'world' };
> 
> // src/index1.ts
> import Shared from './shared';
> console.log('index1.ts', Shared);
> 
> // src/index2.ts
> import Shared from './shared';
> console.log('index2.ts', Shared);
> ```
> 
> `dist/`:
> 
> ```js
> // dist/index1.js
> var Shared = { hello: 'world' };
> console.log('index1.ts', Shared);
> 
> // dist/index2.js
> var Shared = { hello: 'world' };
> console.log('index2.ts', Shared);
> ```
> 
> I hope this helps someone.
> 
> I didn't put too much effort into figuring out if its possible to auto detect the targets.

Firstly, doing this only for specific targets doesn't do much. You practically need to do it for _all_ non-entry modules, thereby separating their dependency tries.
There are issues with this approach that won't manifest on such a small example, but will most definitely manifest on any >= medium-size project. Namely, the approach will exponentially increase memory consumption. 
Remember, rollup keeps all loaded modules in memory.
This is mitigated through caching.
What this is effectively doing is preventing cache.
Not to get me wrong, I'd be very happy if this wasn't the case xd

**Comment by npearson72** [comment]
It's been half a decade since this issue was raised. I think it's safe to say it won't be resolved any time soon.

As someone new to browser extensions, I'm wondering if others here might help clarify if this issue also appears when using Webpack?

Some of the comments suggest that Webpack doesn't have these issues. If this is true, then perhaps using RSBuild in favor of Vite would be the way to go. Can anyone confirm?

**Comment by npearson72** [comment]
No one is criticizing the lack of progress. But it certainly is a worthwhile consideration for those concerned with this issue. We're grateful for Rollup, but given that it lacks a feature needed by many of the people in this thread, perhaps it's time to discuss whether it's the right choice given that this issue has seemingly not been given the attention other would hope over a relatively vast span of time.

PS: Censorship is never the answer

> I will nuke from orbit any further replies that mention progress

**Comment by hood** [comment]
Why was a totally pacific and reasonable comment deleted?

**Comment by Pauan** [comment]
Telling people to "step up and contribute" seems incompatible with also taking an extremely draconian censorship approach. That's more likely to just chase people away, resulting in far fewer contributions.
